src/tests/NXTTest.cpp

```diff
@@ -0,0 +1,277 @@
+            case OpenGLBackend:
+                return utils::BackendType::OpenGL;
+            case VulkanBackend:
+                return utils::BackendType::Vulkan;
+            default:
+                ASSERT(false);
+                return utils::BackendType::Null;
+        }
+    }
+
+    std::string ParamName(BackendType type) {
+        switch(type) {
+            case D3D12Backend:
+                return "D3D12";
+            case MetalBackend:
+                return "Metal";
+            case OpenGLBackend:
+                return "OpenGL";
+            case VulkanBackend:
+                return "Vulkan";
+            default:
+                ASSERT(false);
+                return "";
+        }
+    }
+
+    // Windows don't usually like to be bound to one API than the other, for example switching
+    // from Vulkan to OpenGL causes crashes on some drivers. Because of this, we lazily created
+    // a window for each backing API.
+    GLFWwindow* windows[NumBackendTypes];
+
+    // Creates a GLFW window set up for use with a given backend.
+    GLFWwindow* GetWindowForBackend(utils::BackendBinding* binding, BackendType type) {
+        GLFWwindow** window = &windows[type];
+
+        if (*window != nullptr) {
+            return *window;
+        }
+
+        if (!glfwInit()) {
+            return nullptr;
+        }
+
+        glfwDefaultWindowHints();
+        binding->SetupGLFWWindowHints();
+
+        std::string windowName = "NXT " + ParamName(type) + " test window";
+        *window = glfwCreateWindow(400, 400, windowName.c_str(), nullptr, nullptr);
+
+        return *window;
+    }
+
+    // End2end tests should test valid commands produce the expected result so no error
+    // should happen. Failure cases should be tested in the validation tests.
+    void DeviceErrorCauseTestFailure(const char* message, nxtCallbackUserdata) {
+         FAIL() << "Device level failure: " << message;
+    }
+
+    struct MapReadUserdata {
+        NXTTest* test;
+        size_t slot;
+    };
+}
+
+NXTTest::~NXTTest() {
+    // We need to destroy child objects before the Device
+    readbackSlots.clear();
+    queue = nxt::Queue();
+    device = nxt::Device();
+
+    delete binding;
+    binding = nullptr;
+
+    nxtSetProcs(nullptr);
+}
+
+void NXTTest::SetUp() {
+    binding = utils::CreateBinding(ParamToBackendType(GetParam()));
+    ASSERT(binding != nullptr);
+
+    GLFWwindow* testWindow = GetWindowForBackend(binding, GetParam());
+    ASSERT(testWindow != nullptr);
+
+    binding->SetWindow(testWindow);
+
+    nxtDevice backendDevice;
+    nxtProcTable backendProcs;
+    binding->GetProcAndDevice(&backendProcs, &backendDevice);
+
+    nxtSetProcs(&backendProcs);
+    device = nxt::Device::Acquire(backendDevice);
+    queue = device.CreateQueueBuilder().GetResult();
+
+    device.SetErrorCallback(DeviceErrorCauseTestFailure, 0);
+}
+
+void NXTTest::TearDown() {
+    binding->SwapBuffers();
```

Comment:
This is something we do in ANGLE so that you can see the graphics rendering flashing on the window to try to understand what's going on. Removed.